Docs: Add xnack to unified memory page (#3732)
ROCm-Developer-Tools/HIP: docs/how-to/hip_runtime_api/memory_management/unified_memory.rst

Reviewer: Fir line 13:
space and dedicated physical memory backing it up, for example normal RAM for CPUs and

Author: Fixed

Reviewer: For line 49:
the host or another device, on which the memory is located. The page is then

Author: Fixed

:sup:`1` Works only with ``XNACK=1`` and kernels with HMM support. First GPU
access causes recoverable page-fault.

.. _xnack:

XNACK
-----

> On specific GPU architectures (which is referenced in the previous table), there is an

Reviewer: ```suggestion
On specific GPU architectures (referenced in the previous table), there is an
```

Author: Fixed

option to automatically migrate pages of memory between host and device. This is important
for managed memory, where the locality of the data is important for performance.
Depending on the system, page migration may be disabled by default in which case managed
memory will act like pinned host memory and suffer degraded performance.

> **XNACK** describes the GPU's ability to retry memory accesses that failed due to a page fault

Reviewer: Although it is footnoted in the previous table, I would add here that you need to enable XNACK ability by setting XNACK=1 environment variable.

Author: Added the sentence, what do you think?

(which normally would lead to a memory access error), and instead retrieve the missing page.

This also affects memory allocated by the system as indicated by the first table in
:ref:`unified memory allocators`.

Below is a small example that demonstrates an explicit page fault and how **XNACK** affects
the page fault behavior.

.. code-block:: cpp

    #include <hip/hip_runtime.h>
    #include <iostream>

    #define HIP_CHECK(expression)              \
    {                                          \
        const hipError_t err = expression;     \
        if(err != hipSuccess){                 \
            std::cerr << "HIP error: "         \
                << hipGetErrorString(err)      \
                << " at " << __LINE__ << "\n"; \
        }                                      \

>         exit(EXIT_FAILURE);                    \

Reviewer 2: This currently always exits the program

Author: Oversight on my part, fixed

    }

    __global__ void write_to_memory(int* data, int size)
    {
        int idx = blockIdx.x * blockDim.x + threadIdx.x;
        if (idx < size)
        {
            // Writing to memory that may not have been allocated in GPU memory
            data[idx] = idx * 2; // Triggers a page fault if not resident
        }
    }

    int main()
    {
        const int N = 1024; // 1K elements
        const int blocksize = 256;
        int* data;

        // Allocate unified memory
        HIP_CHECK(hipMallocManaged(&data, N * sizeof(int)));

        // Intentionally don't initialize or prefetch any part of the data
        // No initialization: data is uninitialized but accessible

        // Launch kernel that writes to all elements
        dim3 threads(blocksize);

>         dim3 grids(N / blocksize)

Reviewer 2: This is the amount of blocks, not grids (also missing a `;`)

Author: Fixed

        hipLaunchKernelGGL(write_to_memory, grids, threads, 0, 0, data, N);

>         // Synchronize to ensure kernel execution and fault resolution

Reviewer 2: Still: This does not ensure kernel execution, but kernel completion/termination

Author: Fixed

        HIP_CHECK(hipDeviceSynchronize());

        // Check results
        bool pass = true;

>         for (int i = 0; i < 10; ++i)

Reviewer 2: Please verify the whole array

Author: Fixed

                break;
            }
        }

        if (pass)
        {
            std::cout << "Passed" << std::endl;
        }

        // Free memory
        HIP_CHECK(hipFree(data));
        return 0;
    }

The key behaviors in the example above are as follows:-

#. | No Prefetch or Initialization: The memory is allocated using hipMallocManaged, but
   | it's not initialized or explicitly prefetched to the GPU.

#. | Kernel Write: The kernel writes to the entire array, including memory locations
   | that haven't been allocated in GPU memory yet. This triggers page faults for pages
   | not currently mapped to the GPU.

#. | If **XNACK** is enabled, page faults are handled gracefully: the runtime allocates
   | or fetches the missing pages as needed, ensuring correct execution. If **XNACK** is
   | disabled, the GPU would not handle the page faults, leading to undefined behavior.

To check if page migration is available on a platform, use ``rocminfo``:

.. code-block:: bash

    $ rocminfo | grep xnack
    Name:                    amdgcn-amd-amdhsa--gfx90a:sramecc+:xnack-

Here, ``xnack-`` means that XNACK is available but is disabled by default.

> Turning on XNACK by setting the environment variable ``HSA_XNACK=1`` gives

Reviewer: Here it says that XNACK is enabled with HSA_XNACK=1, but up above it is simply XNACK=1. Which is correct?

Author: I checked, and it would appear that HSA_XNACK is the correct variable. I saw another issue which highlighted this confusion too. Fixed it

        making an explicit copy, like a normal memory access, hence the term
        "zero copy".

.. _xnack:

XNACK
-----

> On a subset of GPUs, such as the MI200, there is an option to automatically migrate pages

Reviewer 2: Can this subset be specified closer than just "such as the MI200"?

of memory between host and device. This is important for managed memory, where the locality
of the data is important for performance.
Depending on the system, page migration may be disabled by default in which case managed
memory will act like pinned host memory and suffer degraded performance.

> **XNACK** describes the GPUs ability to retry memory accesses that failed due a page fault

Reviewer 2: ```suggestion
**XNACK** describes the GPU's ability to retry memory accesses that failed due to a page fault
```

> **XNACK** describes the GPUs ability to retry memory accesses that failed due a page fault

Reviewer 2: ```suggestion
**XNACK** describes the GPU's ability to retry memory accesses that failed due to a page fault
```

(which normally would lead to a memory access error), and instead retrieve the missing page.

This also affects memory allocated by the system as indicated by the previous table.

Below is a small example that demonstrates an explicit page fault and how **XNACK** affects
the page fault behavior.

.. code-block:: cpp

    #include <hip/hip_runtime.h>
    #include <iostream>

    #define HIP_CHECK(expression)              \
    {                                          \
        const hipError_t err = expression;     \
        if(err != hipSuccess){                 \
            std::cerr << "HIP error: "         \
                << hipGetErrorString(err)      \
                << " at " << __LINE__ << "\n"; \

>         }                                      \

Reviewer 2: Just noticed this, and a lot of other example code suffers from this: this continues execution after an error has been encountered. Should be tackled at some point, not necessarily in this PR.

    }

    __global__ void write_to_memory(int* data, int size)
    {
        int idx = blockIdx.x * blockDim.x + threadIdx.x;
        if (idx < size)
        {
            // Writing to memory that may not have been allocated in GPU memory
            data[idx] = idx * 2; // Triggers a page fault if not resident
        }
    }

    int main()
    {
        const int N = 1024; // 1K elements
        int* data;

        // Allocate unified memory
        HIP_CHECK(hipMallocManaged(&data, N * sizeof(int)));

        // Intentionally don't initialize or prefetch any part of the data
        // No initialization: data is uninitialized but accessible

        // Launch kernel that writes to all elements

>         hipLaunchKernelGGL(write_to_memory, dim3(4), dim3(256), 0, 0, data, N);

Reviewer 2: No magic numbers. The `4` in here is `N/256`, or rather `blocks = N/blocksize`

```suggestion
        hipLaunchKernelGGL(write_to_memory, blocks, blocksize, 0, 0, data, N);
```

>         // Synchronize to ensure kernel execution and fault resolution

Reviewer 2: Slightly misleading, synchronization is not needed for kernel execution, but ensuring that it finished executing

        HIP_CHECK(hipDeviceSynchronize());

        // Print a few results
        for (int i = 0; i < 10; ++i)
        {
            std::cout << "data[" << i << "] = " << data[i] << std::endl;

>         }

Reviewer 2: Why not verify all results and just print "passed" or errors if any.

        // Free memory
        HIP_CHECK(hipFree(data));
        return 0;
    }

The key behaviors in the example above are as follows:-

#. | No Prefetch or Initialization: The memory is allocated using hipMallocManaged, but
   | it's not initialized or explicitly prefetched to the GPU.

#. | Kernel Write: The kernel writes to the entire array, including memory locations
   | that haven't been allocated in GPU memory yet. This triggers page faults for pages
   | not currently mapped to the GPU.

#. | If **XNACK** is enabled, page faults are handled gracefully: the runtime allocates
   | or fetches the missing pages as needed, ensuring correct execution. If **XNACK** is
   | disabled, the GPU would not handle the page faults, leading to undefined behavior.

To check if page migration is available on a platform, use ``rocminfo``:

.. code-block:: bash

    $ rocminfo | grep xnack
    Name:                    amdgcn-amd-amdhsa--gfx90a:sramecc+:xnack-

Here, ``xnack-`` means that XNACK is available but is disabled by default.

> Turning on XNACK by setting the environment variable ``HSA_XNACK=1`` and gives

Reviewer 2: ```suggestion
Turning on XNACK by setting the environment variable ``HSA_XNACK=1`` gives
```

the expected result, ``xnack+``:

.. code-block:: bash

    $ HSA_XNACK=1 rocminfo | grep xnack
    Name:                    amdgcn-amd-amdhsa--gfx90a:sramecc+:xnack+

> ``hipcc`` by default will generate code that runs correctly with both XNACK enabled or disabled.

Reviewer 2: Don't mix up tenses :)
```suggestion
``hipcc`` by default generates code that runs correctly with both XNACK enabled or disabled.
```

Setting the ``--offload-arch=``-option with ``xnack+`` or ``xnack-`` forces code to
be only run with XNACK enabled or disabled respectively.

.. code-block:: bash

    # Compiled kernels will run regardless if XNACK is enabled or is disabled.
    hipcc --offload-arch=gfx90a

    # Compiled kernels will only be run if XNACK is enabled with XNACK=1.
    hipcc --offload-arch=gfx90a:xnack+

    # Compiled kernels will only be run if XNACK is disabled with XNACK=0.

>     hipcc --offload-arch=gfx90a:xnack-

Reviewer 2: The wording here sounds weird. Isn't it the case, that if a program relies on functions with these, but they are only compiled for one case, then it will throw an error along the lines of "no kernel for this device available" or something like that? So it's not just that the kernel "will not be run", but they will actually fail to execute
